Gizmo: Type Assert All Values

**Description**
Currently, Gizmo will silently fail on values which are incorrectly formatted strings or hard fail with a weird error message for incorrect types. To reduce confusion values should be validated before execution.

Example of panic caused because of not type checking:
```javascript
g.V().filter(/hello/).all()
```
Error:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: Panic at 10: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: Panic at 10: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x88ea45]

goroutine 13 [running]:
github.com/dop251/goja.(*Runtime).RunProgram.func1(0xc0007a9e58)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/runtime.go:880 +0x9b
panic(0x13751e0, 0xc02a476320)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/dop251/goja.(*vm).try.func1(0xc00027c000, 0x0, 0xc0007a9d58, 0x0, 0x0, 0x0, 0xc0007a9db0)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/vm.go:364 +0x459
panic(0x13932a0, 0x2178890)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cayleygraph/cayley/graph/shape.Filter.BuildIterator(0x1764f20, 0x233ff28, 0xc02a476260, 0x1, 0x1, 0x177af00, 0xc00028c310, 0x2, 0xc00028c300)
        /cayley/graph/shape/shape.go:305 +0xb5
github.com/cayleygraph/cayley/graph/shape.BuildIterator(0x177af00, 0xc00028c310, 0x1764fe0, 0xc016855ce0, 0xc02a225400, 0x0)
        /cayley/graph/shape/shape.go:221 +0x21b
github.com/cayleygraph/cayley/graph/path.(*Path).BuildIteratorOn(0xc027adb040, 0x177af00, 0xc00028c310, 0xc0400a8e00, 0x4)
        /cayley/graph/path/path.go:513 +0x57
github.com/cayleygraph/cayley/query/gizmo.(*pathObject).buildIteratorTree(0xc02a225120, 0x20, 0xc019db7860)
        /cayley/query/gizmo/traversals.go:84 +0x4c
github.com/cayleygraph/cayley/query/gizmo.(*pathObject).GetLimit(0xc02a225120, 0x64, 0x472d96, 0x1319340)
        /cayley/query/gizmo/finals.go:28 +0x2f
github.com/cayleygraph/cayley/query/gizmo.(*pathObject).All(0xc02a225120, 0x0, 0x0)
        /cayley/query/gizmo/finals.go:37 +0x37
reflect.callMethod(0xc027adb100, 0xc0007a96b0, 0xc0007a9698)
        /usr/local/go/src/reflect/value.go:689 +0x180
reflect.methodValueCall(0x0, 0x0, 0xc03cd12501, 0xc02a476300, 0xc0007a9998, 0x4b7454, 0xc019db7710, 0xc027adb100, 0xc02a476300, 0x10, ...)
        /usr/local/go/src/reflect/asm_amd64.s:35 +0x42
reflect.Value.call(0x1311180, 0xc027adb100, 0x13, 0x151f186, 0x4, 0x233ff28, 0x0, 0x0, 0x1517480, 0x1, ...)
        /usr/local/go/src/reflect/value.go:447 +0x454
reflect.Value.Call(0x1311180, 0xc027adb100, 0x13, 0x233ff28, 0x0, 0x0, 0x1065451, 0xc02a225320, 0x203005)
        /usr/local/go/src/reflect/value.go:308 +0xa4
github.com/dop251/goja.(*Runtime).wrapReflectFunc.func1(0x177d660, 0xc02a225160, 0xc0000ab8e0, 0x0, 0x4, 0x0, 0x0)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/runtime.go:1153 +0x7c3
github.com/dop251/goja.(*vm)._nativeCall(0xc00027c000, 0xc040011760, 0x0)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/vm.go:1826 +0x28d
github.com/dop251/goja.call.exec(0x0, 0xc00027c000)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/vm.go:1810 +0x4e3
github.com/dop251/goja.(*vm).run(0xc00027c000)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/vm.go:288 +0x5d
github.com/dop251/goja.(*vm).run-fm()
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/vm.go:375 +0x2a
github.com/dop251/goja.(*vm).try(0xc00027c000, 0xc0007a9db8, 0x0)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/vm.go:370 +0x10c
github.com/dop251/goja.(*vm).runTry(0xc00027c000, 0x15768c0)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/vm.go:375 +0x4e
github.com/dop251/goja.(*Runtime).RunProgram(0xc000479200, 0xc000150c60, 0x0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/dop251/goja@v0.0.0-20190105122144-6d5bf35058fa/runtime.go:891 +0x1b7
github.com/cayleygraph/cayley/query/gizmo.(*Session).run(0xc0001326c0, 0xc028bfa160, 0x1e, 0xc000150120, 0xc0001326c0, 0xc01987c0c0, 0x0)
        /cayley/query/gizmo/gizmo.go:272 +0xb4
github.com/cayleygraph/cayley/query/gizmo.(*Session).Execute(0xc0001326c0, 0x176fce0, 0xc000150120, 0xc028bfa160, 0x1e, 0xc000150ba0, 0x64)
        /cayley/query/gizmo/gizmo.go:295 +0x172
created by github.com/cayleygraph/cayley/internal/http.(*API).ServeV1Query
        /cayley/internal/http/query.go:126 +0x45f
```

--- a/query/gizmo/gizmo_test.go
+++ b/query/gizmo/gizmo_test.go
@@ -194,6 +194,13 @@ var testQueries = []struct {
 		expect: []string{"<charlie>"},
 	},
 	{
+		message: "filter with a wrong type",
+		query: `
+			g.V().filter(/<alice>/).all()
+		`,
+		err: true,
+	},
+	{
 		message: "use .both()",
 		query: `
 			g.V("<fred>").both("<follows>").all()

--- a/query/gizmo/traversals.go
+++ b/query/gizmo/traversals.go
@@ -17,6 +17,7 @@ package gizmo
 // Adds special traversal functions to JS Gizmo objects. Most of these just build the chain of objects, and won't often need the session.
 
 import (
+	"errors"
 	"fmt"
 
 	"github.com/dop251/goja"
@@ -644,6 +645,9 @@ func (p *pathObject) Filter(args ...valFilter) (*pathObject, error) {
 	}
 	filt := make([]shape.ValueFilter, 0, len(args))
 	for _, f := range args {
+		if f.f == nil {
+			return nil, errors.New("invalid argument type in filter()")
+		}
 		filt = append(filt, f.f)
 	}
 	np := p.clonePath().Filters(filt...)

